packages/engine-server/src/markdown/remark/utils.ts

```diff
@@ -179,6 +179,47 @@ const getLinks = ({
   return dlinks;
 };
 
+const getUnreferencedLinks = ({
+  ast,
+  note,
+  notes,
+}: {
+  ast: DendronASTNode;
+  note: NoteProps;
+  notes: NoteProps[];
+}) => {
+  const textNodes: Node[] = [];
+  visit(ast, [DendronASTTypes.TEXT], (node) => {
+    switch (node.type) {
+      case DendronASTTypes.TEXT:
+        textNodes.push(node);
+    }
+  });
+  const unreferencedLinks: DLink[] = [];
+
+  _.forEach(notes, (n) => {
+    const fname = n.fname;
```

Comment:
doesn't need to be implemented now but we might be able to optimize in the future by creating a massive regex 

```ts
// pseudo code
// take 100 notes
notes.batch(100).map(batch => {
	re = RegExp(batch.notes.map(n => n.fname).join("|"))
	re.matchAll(text)
	...
})
```

Reply:
Another option would be creating a prefix-tree data structure with the `fname`s: https://en.wikipedia.org/wiki/Trie

Reply:
Also, here's a simple but more efficient way to implement this:

1. Create a map of `fname`s to notes
2. After getting all the `Text` nodes from a note, split the text into words, and test if each word appears in the map.

This would have `O(n.logn)` running time, as opposed to the quadratic `O(n^2)` running time of this implementation. The only drawback is that this assumes `fname`s can't contain spaces.

It is possible to allow this to work with `fname`s containing spaces too: use the first word of the `fname` as the map key, and the list of notes with that word as the first word of their `fname` as the map value. When parsing for unlinked references, if a lookup in this map is true, then search through the list in the value for matches. This way you don't have to search through every existing `fname`, only a few ones that have the same word as the first word of their `fname`.

Reply:
Thanks for the feedback. I went with @SeriousBug 's second suggestion.

I definitely need to break out of my sloppy habits and embrace FP 😅